### PR TITLE
Allow dtype input argument in np.sum

### LINKED ISSUE #4220: support for 'dtype' keyword argument in np.sum()
## Feature request: Add `dtype` keyword argument to `np.sum()`

I am working on a scientific toolbox that relies on numpy and scipy and requires decent numeric precision. To parallelize core parts, I considered using Numba as an _optional dependency_ with a fallback to native numpy. This seems doable and elegant except for the following:

The existing code relies heavily on `np.mean()` with the `axis` and `dtype` keyword arguments. `dtype=np.float64` is needed to get enough precision in certain situations.

* This does _not_ cast (copy to new dtype) the input array, only the temporary variable is of a higher precision.
* I managed to implement this in numba with a manual for loop, casting _within_ the loop
* The problem arises when using arrays of higher dimension. A specific function has to be written for every ndim-to-axis combination. As far as i understand, one cannot tell numba to check if-conditions for dimensionality during compilation, to avoid `tuple index out of range`.

The script below illustrates the precision issue:

```
import numpy as np
import numpy.random
import time

try:
    from numba import jit, njit, prange, gdb
    # raise ImportError # uncomment to toggle numba
except ImportError:
    print('Numba not available, skipping parallelization')
    # replace numba functions if numba not available:
    # we only use jit and prange
    # helper needed for decorators with kwargs
    def parametrized(dec):
        def layer(*args, **kwargs):
            def repl(f):
                return dec(f, *args, **kwargs)
            return repl
        return layer

    @parametrized
    def jit(func, **kwargs):
        return func
    def prange(*args):
        return range(*args)

ftype = np.float64


@jit(nopython=True, parallel=False, fastmath=True)
def mean_lowprec(a):
    return np.sum(a)/(2*a.shape[1])

@jit(nopython=True, parallel=False, fastmath=True)
def mean_copy(a):
    # copies in ram, factor two memory required
    b = a.astype(ftype)
    return np.sum(b)/(2*a.shape[1])

@jit(nopython=True, parallel=False, fastmath=True)
def mean_2d_reduce(a):
    total = ftype(0)
    for i in range(a.shape[0]):
        for j in range(a.shape[1]):
            total += ftype(a[i,j])

    return total/ftype(a.shape[0]*a.shape[1])


a = np.zeros((2, 512*512), dtype=np.float32)
a[0, :] = 1.0
a[1, :] = 0.1

# compile first call
mean_lowprec(a)
mean_copy(a)
mean_2d_reduce(a)

time_beg = time.time()
print(f'np.sum f4:           {np.sum(a)/(2*a.shape[1])}')
print(f'{(time.time()-time_beg)*1000 :.2f} ms\n')

time_beg = time.time()
print(f'np.sum f8:           {np.sum(a, dtype=ftype)/(2*a.shape[1])}')
print(f'{(time.time()-time_beg)*1000 :.2f} ms\n')

time_beg = time.time()
print(f'mean_lowprec(a):     {mean_lowprec(a)}')
print(f'{(time.time()-time_beg)*1000 :.2f} ms\n')

time_beg = time.time()
print(f'mean_copy(a):        {mean_copy(a)}')
print(f'{(time.time()-time_beg)*1000 :.2f} ms\n')

time_beg = time.time()
print(f'mean_2d_reduce(a):   {mean_2d_reduce(a)}')
print(f'{(time.time()-time_beg)*1000 :.2f} ms\n')
```

Results in
```
np.sum f4:           0.5499992370605469
0.49 ms

np.sum f8:           0.5500000007450581
0.55 ms

mean_lowprec(a):     0.5498046875
0.10 ms

mean_copy(a):        0.5500000007450581
0.84 ms

mean_2d_reduce(a):   0.5500000007450581
0.15 ms
```

While the last option seems a good candidate performance/memory wise, I failed to come up with a solution that offers a one-wrapper-function fits-all to mimic the native `numpy.mean()`.

Hence, I believe support for the `dtype` argument in numba would be a great feature!
I would greatly appreciate ideas for a workaround or some background on the details I am missing. 

Thanks,
Paul

### PR DESCRIPTION
**Pull request by luk-f-a**
closes #4220

### CODE REVIEW

**Comment by esc** [comment]
@luk-f-a thanks for getting this going! I'll mark it as `in-progress` for now, please do let us know when it is ready for a review.

**Comment by luk-f-a** [comment]
I have implemented the case `np.sum(arr, dtype)`. Before I move on to `np.sum(arr, axis, dtype)` it would be helpful if someone could perform a quick review and tell me if I'm on the right path.

**Comment by esc** [comment]
@luk-f-a thanks, I have scheduled it for a preliminary review.

**Comment by stuartarchibald** [comment]
@luk-f-a thanks for the patch. I think this approach is likely to work, but as noted on gitter, have to wonder about the longevity of such an addition. That said, it seems to be less involved than anticipated, so perhaps complete this and ensure there's an exceptionally good set of tests for it and this can then really help guide refactoring these implementations to use the newer APIs at some point in the future. What do you think?

**Comment by luk-f-a** [comment]
@stuartarchibald, sounds good, I'll add more tests. Do you think this will be replaced in the short term?

**Comment by luk-f-a** [comment]
@stuartarchibald, I have added more tests. Maybe too many :smile: . I test over a full range of dtypes, for all 4 implentations: `sum`, `sum(axis)`, `sum(dtype)` and `sum(axis, dtype)`. some are failing for certain dtypes.
-`sum` passes for all dtypes.
-`sum(axis)` fails for `timedelta`
-`sum(dtype)` passes for all (shares implementation with `sum`)
-`sum(axis, dtype)` (shares implementation with `sum(axis)`) fails for `timedelta` and `int32`.

The error  with timedelta comes from `type(zero)` in `arraymath.py` line 203. do you know why this would fail?
The error with int32 happens because the result is returned as int64, I haven't been able to track this yet.
Please have a look and let me know if these tests are sufficient for what you had in mind.

**Comment by stuartarchibald** [comment]
@luk-f-a, thanks for this, I've taken a look at the tests, this level of detail seems appropriate given the changes proposed. I think some more effort will be needed code generation wise as it seems that the compiler might be being hit too often, will take a closer look when this is nearer ready. 

> The error with timedelta comes from type(zero) in arraymath.py line 203. do you know why this would fail?

Suspect the timedelta types are lacking `__class__` attr resolvers cf.:
https://github.com/numba/numba/blob/60d2bddd0a9ddf84cf900a70e4b7e17bcbe5dc87/numba/typing/builtins.py#L681-L704

**Comment by luk-f-a** [comment]
@stuartarchibald , I've fixed the problem with `np.int32`. All that remains is the test failing on sums of `timedelta` arrays. What is failing is the existing `np.sum` implementation, not the new code, should I open a separate issue to fix `timedelta` and disable that test for now?
I tried to fix it, but I cannot find the right place. You showed me an example of `NumberAttribute` template, but there's nothing similar for `timedelta`. Should one be created?

**Comment by stuartarchibald** [comment]
@luk-f-a excellent, good job. Yes, if the `timedelta` etc problems are present in existing implementations then please file a new issue for them and feel free to ignore here. I think that there's sufficient quantity of churn/new code here without trying to fix bugs/unimplemented corners too!

**Comment by stuartarchibald** [comment]
Seems like there's some windows CI failures, suspect it's `types.intp`/windows having different integer size related.

**Comment by luk-f-a** [comment]
@stuartarchibald I created an issue for `timedelta`. Regarding windows, I am trying to set up a Windows env to test locally. Have you seen this problem before? What's the usual approach when linux and Windows behave differently?

**Comment by stuartarchibald** [comment]
> @stuartarchibald I created an issue for `timedelta`. Regarding windows, I am trying to set up a Windows env to test locally. Have you seen this problem before? What's the usual approach when linux and Windows behave differently?

Yes, this is a really common issue. It's usually just `intp` related and unfortunately often requires a windows box to debug.

**Comment by luk-f-a** [comment]
@stuartarchibald , I set up the Windows env. It turns out that Numba returns consistent dtypes across Windows and Linux (`uint32` is always cast to `uint64` for `np.sum`) but Numpy casts to `uint64` on Linux and keeps `uint32` on Linux. ¯\_(ツ)_/¯
How far do you want to take the idea that Numba should return the same as Numpy? Personally, I would not try to replicate this Numpy behaviour. In this case (`uint32`), I would change the test to ensure the contents are the same but allowing the dtype to be different in Numpy vs Numba.

**Comment by stuartarchibald** [comment]
> @stuartarchibald , I set up the Windows env. It turns out that Numba returns consistent dtypes across Windows and Linux (`uint32` is always cast to `uint64` for `np.sum`) but Numpy casts to `uint64` on Linux and keeps `uint32` on Linux. ¯_(ツ)_/¯

Do you mean NumPy uses `uint64` for the unsigned accumulator on Linux and `uint32` for the unsigned accumulator on windows (assuming a `uint{8, 16, 32})` input on both)?

> How far do you want to take the idea that Numba should return the same as Numpy? Personally, I would not try to replicate this Numpy behaviour. In this case (`uint32`), I would change the test to ensure the contents are the same but allowing the dtype to be different in Numpy vs Numba.

I think we should get to the bottom of this as summation is such an important operation, it'll help with future work on moving to pairwise summation/writing a summation factory.

**Comment by luk-f-a** [comment]
> Do you mean NumPy uses uint64 for the unsigned accumulator on Linux and uint32 for the unsigned accumulator on windows (assuming a uint{8, 16, 32}) input on both)?


if I run `np.sum(np.ones(10, dtype=np.uint32)).dtype` on Linux (ubuntu 64bits, numpy 1.16.4) I get `dtype('uint64')`. On Windows (win10, numpy 1.16.4), I get `dtype('uint32')`.

It is probably related to this comment https://github.com/numpy/numpy/issues/7531#issuecomment-466053990

**Comment by stuartarchibald** [comment]
Following a bit of discussion with @seibert, we concluded that we should not replicate NumPy and instead continue to use 64bit accumulators and note the difference in documentation.

**Comment by luk-f-a** [comment]
that sounds good. Then that means that I will:
- make a separate test for `uint32` and `int32` where I can check content and not dtype.
- update the documentation to add support for `dtype` parameter and explain the accumulator difference. Also that `dtype` cannot be `datetime`.

anything else comes to mind?

**Comment by stuartarchibald** [comment]
> that sounds good. Then that means that I will:
> 
>     * make a separate test for `uint32` and `int32` where I can check content and not dtype.

Perhaps assert that the dtype is what is expected for Numba?

>     * update the documentation to add support for `dtype` parameter and explain the accumulator difference. Also that `dtype` cannot be `datetime`.

Sounds good.

> anything else comes to mind?

Nothing immediate from this discussion. The PR needs to go through code review next :)

Thanks.

**Comment by stuartarchibald** [comment]
@luk-f-a there's a couple of conflicts to resolve please, preferably by merge of master into this branch, thanks.

**Comment by luk-f-a** [comment]
@stuartarchibald , yes please, I think it's ready for review.

File: numba/targets/arraymath.py [hunk 1]
```diff
@@ -268,8 +226,129 @@ def array_sum_impl_axis(arr, axis):
...
+    compiled = register_jitable(gen_impl)
+
+    def array_sum_impl_axis(arr, axis, dtype):
+        return compiled(arr, axis)
+
+    res = context.compile_internal(builder, array_sum_impl_axis, sig, args)
+    return impl_ret_new_ref(context, builder, sig.return_type, res)
+
+
+@lower_builtin(np.sum, types.Array,  types.DTypeSpec)
```
Comment (esc) [inline numba/targets/arraymath.py]: This seems duplicated?

File: numba/typing/arraydecl.py [hunk 2]
```diff
@@ -612,22 +612,38 @@ def sum_expand(self, args, kws):
     sum can be called with or without an axis parameter.
```
Comment (esc) [inline numba/typing/arraydecl.py]: This documentation line should probably be updated to reflect the new signature.

File: numba/typing/npydecl.py [hunk 3]
```diff
@@ -378,9 +378,18 @@ def generic(self, args, kws):
...
         if kws:
             if self.method_name == 'sum':
-                def sum_stub(arr, axis):
-                    pass
-                pysig = utils.pysignature(sum_stub)
+                if 'axis' in kws and not 'dtype' in kws:
+                    def sum_stub(arr, axis):
+                        pass
+                    pysig = utils.pysignature(sum_stub)
+                if 'dtype' in kws and not 'axis' in kws:
```
Comment (esc) [inline numba/typing/npydecl.py]: Same comment as above on the negation, also, this could be an `elif` clause.

File: numba/typing/arraydecl.py [hunk 4]
```diff
@@ -612,22 +612,38 @@ def sum_expand(self, args, kws):
...
     """
     pysig = None
-    if kws:
+    if 'axis' in kws and not 'dtype' in kws:
```
Comment (esc) [inline numba/typing/arraydecl.py]: Prefer to use `not in`.

File: numba/typing/arraydecl.py [hunk 5]
```diff
@@ -612,22 +612,38 @@ def sum_expand(self, args, kws):
...
         def sum_stub(axis):
             pass
         pysig = utils.pysignature(sum_stub)
         # rewrite args
         args = list(args) + [kws['axis']]
-        kws = None
+        # kws = None
```
Comment (esc) [inline numba/typing/arraydecl.py]: dead code?

File: numba/typing/arraydecl.py [hunk 6]
```diff
@@ -612,22 +612,38 @@ def sum_expand(self, args, kws):
...
+
+    if 'dtype' in kws and not 'axis' in kws:
+        def sum_stub(dtype):
+            pass
+        pysig = utils.pysignature(sum_stub)
+        # rewrite args
+        args = list(args) + [kws['dtype']]
+        # kws = None
+
+    if 'dtype' in kws and 'axis' in kws:
```
Comment (esc) [inline numba/typing/arraydecl.py]: Candidate for `elif`.

File: numba/targets/arraymath.py [hunk 7]
```diff
@@ -268,8 +226,129 @@ def array_sum_impl_axis(arr, axis):
...
+    return inner
+
+
+@lower_builtin(np.sum, types.Array, types.intp, types.DTypeSpec)
+@lower_builtin(np.sum, types.Array, types.IntegerLiteral, types.DTypeSpec)
+@lower_builtin("array.sum", types.Array, types.intp, types.DTypeSpec)
+@lower_builtin("array.sum", types.Array, types.IntegerLiteral, types.DTypeSpec)
+def array_sum_axis_dtype(context, builder, sig, args):
+    """
+    The third parameter to gen_index_tuple that generates the indexing
```
Comment (esc) [inline numba/targets/arraymath.py]: This comment doesn't appear to be related to the code around it. I am not sure even if it is still needed. It didn't help me understand the code better.

File: numba/targets/arraymath.py [hunk 8]
```diff
@@ -268,8 +226,129 @@ def array_sum_impl_axis(arr, axis):
...
+    def array_sum_impl_axis(arr, axis, dtype):
+        return compiled(arr, axis)
+
+    res = context.compile_internal(builder, array_sum_impl_axis, sig, args)
+    return impl_ret_new_ref(context, builder, sig.return_type, res)
+
+
+@lower_builtin(np.sum, types.Array,  types.DTypeSpec)
+@lower_builtin(np.sum, types.Array,  types.DTypeSpec)
+@lower_builtin("array.sum", types.Array, types.DTypeSpec)
```
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/targets/arraymath.py]: fixed

File: numba/targets/arraymath.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/targets/arraymath.py]: fixed

File: numba/typing/arraydecl.py (hunk 2 above)
Comment (luk-f-a) [reply numba/typing/arraydecl.py]: fixed

File: numba/typing/npydecl.py [hunk 9]
```diff
@@ -378,9 +378,18 @@ def generic(self, args, kws):
         pysig = None
         if kws:
             if self.method_name == 'sum':
-                def sum_stub(arr, axis):
-                    pass
-                pysig = utils.pysignature(sum_stub)
+                if 'axis' in kws and not 'dtype' in kws:
```
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/typing/npydecl.py]: you're right, changed.

File: numba/typing/npydecl.py (hunk 3 above)
Comment (luk-f-a) [reply numba/typing/npydecl.py]: fixed

File: numba/typing/npydecl.py [hunk 10]
```diff
@@ -378,9 +378,18 @@ def generic(self, args, kws):
...
+                    def sum_stub(arr, axis):
+                        pass
+                    pysig = utils.pysignature(sum_stub)
+                if 'dtype' in kws and not 'axis' in kws:
+                    def sum_stub(arr, dtype):
+                        pass
+                    pysig = utils.pysignature(sum_stub)
+                if 'dtype' in kws and 'axis' in kws:
```
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/typing/npydecl.py]: yes, fixed

File: numba/typing/arraydecl.py (hunk 4 above)
Comment (luk-f-a) [reply numba/typing/arraydecl.py]: done

File: numba/typing/arraydecl.py (hunk 5 above)
Comment (luk-f-a) [reply numba/typing/arraydecl.py]: tbh, I couldn't understand why that was there so left it in case I had to restore it. Since everything is working I can delete it. However, it's that feeling you get when you put together a machine and there's a piece left outside. I'll delete it.

File: numba/typing/arraydecl.py [hunk 11]
```diff
@@ -612,22 +612,38 @@ def sum_expand(self, args, kws):
...
+    if 'axis' in kws and not 'dtype' in kws:
         def sum_stub(axis):
             pass
         pysig = utils.pysignature(sum_stub)
         # rewrite args
         args = list(args) + [kws['axis']]
-        kws = None
+        # kws = None
+
+    if 'dtype' in kws and not 'axis' in kws:
```
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/typing/arraydecl.py]: done

File: numba/typing/arraydecl.py [hunk 12]
```diff
@@ -612,22 +612,38 @@ def sum_expand(self, args, kws):
...
+        def sum_stub(dtype):
+            pass
+        pysig = utils.pysignature(sum_stub)
+        # rewrite args
+        args = list(args) + [kws['dtype']]
+        # kws = None
```
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/typing/arraydecl.py]: deleted

File: numba/targets/arraymath.py (hunk 7 above)
Comment (luk-f-a) [reply numba/targets/arraymath.py]: thanks, there was refactoring and this comment should have moved with `gen_sum_axis_impl`. I did that now.

File: numba/tests/test_array_methods.py [hunk 13]
```diff
@@ -779,18 +787,195 @@ def test_sum(self):
         # OK
         self.assertPreciseEqual(pyfunc(a, 0), cfunc(a, 0))
 
-    def test_sum_kws(self):
-        pyfunc = array_sum_kws
+    def test_sum2(self):
+        pyfunc = array_sum
+        cfunc = jit(nopython=True)(pyfunc)
+        all_dtypes = [np.float64, np.float32, np.int64, np.int32,
```
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/tests/test_array_methods.py]: good one, I didn't even know `complex128` existed. I never work with complex numbers. Added it now.

File: numba/targets/arraymath.py (hunk 7 above)
Comment (esc) [reply numba/targets/arraymath.py]: That is what a good review is for. 😄

**Comment by esc** [comment]
I think it is in good shape now. The only thing I would like to discuss maybe, is if it is possible to reduce the duplication between `array_sum_axis` and `array_sum_axis_dtype`. If not and there are good reasons, that will be fine too.

**Comment by esc** [comment]
As discussed via Gitter, a bit of repetition is fine since the code may be overhauled using `overload` in future.

**Comment by esc** [comment]
Ready to merge from my side. @stuartarchibald should this go through any other buildfarm than public CI?

File: docs/source/reference/numpysupported.rst [hunk 14]
```diff
@@ -210,13 +210,21 @@ The following methods of Numpy arrays are supported:
...
-* :meth:`~numpy.ndarray.sum` (with or without the ``axis`` argument. 
-  ``axis`` only supports ``integer`` values)
+* :meth:`~numpy.ndarray.sum` (with or without the ``axis`` and/or ``dtype`` arguments.)
 
+  * ``axis`` only supports ``integer`` values.
   * If the ``axis`` argument is a compile-time constant, all valid values are supported.
     An out-of-range value will result in a ``LoweringError`` at compile-time.
   * If the ``axis`` argument is not a compile-time constant, only values from 0 to 3 are supported.
     An out-of-range value will result in a runtime exception.
+  * All numeric ``dtypes`` are supported as ``dtype`` parameter. ``timedelta`` arrays can
```
Comment (stuartarchibald) [inline docs/source/reference/numpysupported.rst]: this section all needs line wrapping to 80 chars

File: numba/tests/test_array_methods.py [hunk 15]
```diff
@@ -779,18 +787,198 @@ def test_sum(self):
...
+                        self.assertPreciseEqual(pyfunc(arr, dtype=out_dtype),
+                                                cfunc(arr, dtype=out_dtype))
+
+    def test_sum_dtype_kws_negative(self):
+        pyfunc = array_sum_dtype_kws
+        cfunc = jit(nopython=True)(pyfunc)
+        dtype = np.float64
+        # OK
+        a = np.ones((7, 6, 5, 4, 3))
+        self.assertFalse(type(pyfunc(a, dtype=np.int32)) == cfunc(a, dtype=dtype))
```
Comment (stuartarchibald) [inline numba/tests/test_array_methods.py]: This is comparing a type to a value, what is the intention here?

File: numba/tests/test_array_methods.py [hunk 16]
```diff
@@ -779,18 +787,198 @@ def test_sum(self):
         # OK
         self.assertPreciseEqual(pyfunc(a, 0), cfunc(a, 0))
 
-    def test_sum_kws(self):
-        pyfunc = array_sum_kws
+    def test_sum2(self):
```
Comment (stuartarchibald) [inline numba/tests/test_array_methods.py]: It's be really helpful for future refactoring if the intent of the test were described as a docstring/comment. Some of the new tests have this and they've been very helpful.

File: numba/tests/test_array_methods.py [hunk 17]
```diff
@@ -779,18 +787,198 @@ def test_sum(self):
...
+             np.ones((7, 3), arr_dtype) * -5]
+            for arr_dtype in all_dtypes]
+
+        out_dtypes = {np.dtype('float64'): [np.float64],
+                      np.dtype('float32'): [np.float64, np.float32],
+                      np.dtype('int64'): [np.float64, np.int64, np.float32],
+                      np.dtype('int32'): [np.float64, np.int64, np.float32, np.int32],
+                      np.dtype('uint32'): [np.float64, np.int64, np.float32],
+                      np.dtype('uint64'): [np.float64, np.uint64],
+                      np.dtype('complex64'): [np.complex64],
```
Comment (stuartarchibald) [inline numba/tests/test_array_methods.py]: assume this could also accumulate into a complex128?

File: numba/tests/test_array_methods.py [hunk 18]
```diff
@@ -779,18 +787,198 @@ def test_sum(self):
...
+                                       .format(arr.dtype, out_dtype))
+                        with self.subTest(subtest_str):
+                            py_res = pyfunc(arr, axis=axis, dtype=out_dtype)
+                            nb_res = cfunc(arr, axis=axis, dtype=out_dtype)
+                            self.assertPreciseEqual(py_res, nb_res)
+
+    def test_sum_axis_dtype_pos_arg(self):
+        """
+        testing that axis and dtype inputs work when passed as positional
+        :return:
```
Comment (stuartarchibald) [inline numba/tests/test_array_methods.py]: delete `:return:` ?

File: numba/tests/test_array_methods.py [hunk 19]
```diff
@@ -779,18 +787,198 @@ def test_sum(self):
...
+        cfunc = jit(nopython=True)(pyfunc)
+        dtype = np.float64
+        # OK
+        a = np.ones((7, 6, 5, 4, 3))
+        self.assertPreciseEqual(pyfunc(a, 1, dtype),
+                                cfunc(a,  1, dtype))
+
+        self.assertPreciseEqual(pyfunc(a, 2, dtype),
+                                cfunc(a, 2, dtype))
+
```
Comment (stuartarchibald) [inline numba/tests/test_array_methods.py]: This seems to not work:
```python
from numba import njit
import numpy as np

@njit
def foo(a, dtype):
    z =  a.sum(dtype=dtype, axis=1)
    return z


a = np.ones((2, 3), dtype=np.int8)
print(foo(a, np.uint32))
```

Comment (stuartarchibald) [reply numba/tests/test_array_methods.py]: I'm actually wondering if this should work at all under Numba semantics for `+=`?

File: docs/source/reference/numpysupported.rst (hunk 14 above)
Comment (luk-f-a) [reply docs/source/reference/numpysupported.rst]: done

File: numba/tests/test_array_methods.py (hunk 15 above)
Comment (luk-f-a) [reply numba/tests/test_array_methods.py]: I think the intention (with a mistake) was to compare types but it must have been an early test. the main tests are much better now, this test does not add anything. I'm deleting it.

File: numba/tests/test_array_methods.py (hunk 16 above)
Comment (luk-f-a) [reply numba/tests/test_array_methods.py]: good idea, done

File: numba/tests/test_array_methods.py (hunk 17 above)
Comment (luk-f-a) [reply numba/tests/test_array_methods.py]: yes, added.

File: numba/tests/test_array_methods.py (hunk 18 above)
Comment (luk-f-a) [reply numba/tests/test_array_methods.py]: fixed

File: numba/tests/test_array_methods.py (hunk 19 above)
Comment (luk-f-a) [reply numba/tests/test_array_methods.py]: > This seems to not work:

it's rejecting the transformation of signed integer into unsigned integer. isn't that ok?

> I'm actually wondering if this should work at all under Numba semantics for `+=`?

I didn't understand why `+=` would play a role here.

**Comment by luk-f-a** [comment]
@stuartarchibald  ,`test_sum`, `test_sum_axis_kws` are the original ones, the ones I found in the code when I started. I haven't deleted them but I think the new ones supersede them, since they test the same and more. should I delete them?

File: numba/tests/test_array_methods.py (hunk 19 above)
Comment (stuartarchibald) [reply numba/tests/test_array_methods.py]: I think we can just leave it for now, the complexity of untangling this outweighs the benefit given it will be rewritten/refactored in the near future. The `+=` is to do with reducing where the accumulator is an array https://github.com/numba/numba/pull/4472/files#diff-4b9c6924f184d92f076cc50b84bcc442R224

File: numba/tests/test_array_methods.py (hunk 15 above)
Comment (stuartarchibald) [reply numba/tests/test_array_methods.py]: thanks

File: docs/source/reference/numpysupported.rst (hunk 14 above)
Comment (stuartarchibald) [reply docs/source/reference/numpysupported.rst]: err no :) I can go fix it though so don't worry!

**Comment by stuartarchibald** [comment]
> @stuartarchibald ,`test_sum`, `test_sum_axis_kws` are the original ones, the ones I found in the code when I started. I haven't deleted them but I think the new ones supersede them, since they test the same and more. should I delete them?

If they are entirely superfluous given the new tests then yes please, the build is starting to hit CI wall time limits!

File: docs/source/reference/numpysupported.rst (hunk 14 above)
Comment (luk-f-a) [reply docs/source/reference/numpysupported.rst]: lol, I didn't know it had to be strictly 80. I thought it was like a pep8 thing, where there's a bit of flexibility.

**Comment by stuartarchibald** [comment]
@luk-f-a 0726ed1 seems to have deleted the entire module containing the tests?

**Comment by stuartarchibald** [comment]
Also, https://github.com/luk-f-a/numba/pull/2 word wraps the text.

**Comment by luk-f-a** [comment]
> Also, [luk-f-a#2](https://github.com/luk-f-a/numba/pull/2) word wraps the text.

thanks. 

> seems to have deleted the entire module containing the tests?

sorry, I don't understand how that happened, locally the file was still there. maybe I should not push in the evening.

File: numba/tests/test_array_methods.py [hunk 20]
```diff
@@ -170,6 +179,12 @@ def array_sum_const_axis_neg_one(a, axis):
     # "const_multi" variant would raise errors
     return a.sum(axis=-1)
 
+def array_mean(a):
```
Comment (stuartarchibald) [inline numba/tests/test_array_methods.py]: not sure these are supposed to be here either?

Comment (luk-f-a) [reply numba/tests/test_array_methods.py]: you're too fast, I was just reverting those. I think I got on the wrong side of PyCharm smart checkout, and it brought those from my branch working on `np.mean`. I think I cleaned it now.

**Comment by esc** [comment]
🎉

**Comment by luk-f-a** [comment]
@stuartarchibald , thanks for all your help and patience. I hope this will allow the improvement of other functions (like np.mean getting an `axis` parameter).

**Comment by luk-f-a** [comment]
@esc, thanks for your review and comments.

**Comment by esc** [comment]
@luk-f-a no problem, how we just need to massage it through CI..

**Comment by luk-f-a** [comment]
@esc @stuartarchibald I don't think `Linux py35_np111_cov` has run even once since I added the extended set of tests. It always times out. If you merge this PR, every future PR will have the same problem, won't it?

**Comment by esc** [comment]
@luk-f-a I don't think the cov timeout is caused by this PR.. Think it is pretty safe to assume this PR will be O.K.

**Comment by esc** [comment]
For azure only the coverage run timed out. Probably safe to merge.
